scripts/deep_learning/hierarchy/LaterTier/TrainLaterTierModel_showers.ipynb

```diff
@@ -0,0 +1,897 @@
+    "    loss_func = torch.nn.CrossEntropyLoss(reduction='none')\n",
+    "    loss = loss_func(pred, target)    \n",
+    "    loss = loss * weights\n",
+    "    loss = torch.sum(loss) / loss.shape[0]\n",
+    "    \n",
+    "    return loss\n",
+    "\n",
+    "# Implement own weighting\n",
+    "def loss_function_classifier(pred, target, true_gen, weight_dict) :\n",
+    "\n",
+    "    # I have to reshape so that I can do the weighting - sad.\n",
+    "    pred = pred.reshape(-1)\n",
+    "    target = target.reshape(-1)\n",
+    "    \n",
+    "    # Calculate weights\n",
+    "    weights = torch.ones(true_gen.shape)\n",
+    "    weights[torch.logical_and(target == 1, true_gen == 3)] = weight_dict['secondary_child_true_links_train_weight']\n",
+    "    weights[torch.logical_and(target == 0, true_gen == 3)] = weight_dict['secondary_child_false_links_train_weight']\n",
+    "    weights[torch.logical_and(target == 1, true_gen > 3)] = weight_dict['higher_child_true_links_train_weight']\n",
+    "    weights[torch.logical_and(target == 0, true_gen > 3)] = weight_dict['higher_child_false_links_train_weight']    \n",
+    "    \n",
+    "    # Annoyingly we have to change our target to a float\n",
+    "    target = target.to(torch.float32)    \n",
+    "    \n",
+    "    # Use BCE loss\n",
+    "    loss_func = torch.nn.BCELoss(weight=weights)\n",
+    "    \n",
+    "    # Calculate loss\n",
+    "    loss = loss_func(pred, target)    \n",
+    "\n",
+    "    return loss\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "ae7c4482",
+   "metadata": {
+    "scrolled": false
+   },
+   "outputs": [],
+   "source": [
+    "######################\n",
+    "# Training and testing\n",
+    "######################\n",
+    "\n",
+    "# Optimiser\n",
+    "optimiser = torch.optim.Adam(itertools.chain(model_branch.parameters(), classifier_model.parameters()), lr=LEARNING_RATE)\n",
+    "#loss_func = torch.nn.BCEWithLogitsLoss(pos_weight=class_weights)\n",
+    "\n",
+    "# Scheduler\n",
+    "# scheduler = torch.optim.lr_scheduler.ReduceLROnPlateau(optimiser, 'min', factor=0.1, patience=3, threshold=0.0001, threshold_mode='abs')\n",
+    "\n",
+    "# Put here some metrics\n",
+    "training_epoch = []\n",
+    "training_loss_0 = []\n",
+    "training_loss_1 = []\n",
+    "training_classification_loss = []\n",
+    "training_accuracy = []\n",
+    "training_positive_as_positive_rate = []\n",
+    "training_positive_as_negative_rate = []\n",
+    "training_negative_as_negative_rate = []\n",
+    "training_negative_as_positive_rate = []\n",
+    "\n",
+    "testing_epoch = []\n",
+    "testing_loss_0 = []\n",
+    "testing_loss_1 = []\n",
+    "testing_classification_loss = []\n",
+    "testing_accuracy = []\n",
+    "testing_positive_as_positive_rate = []\n",
+    "testing_positive_as_negative_rate = []\n",
+    "testing_negative_as_negative_rate = []\n",
+    "testing_negative_as_positive_rate = []\n",
```

Comment:
Given there are a lot of metrics here (and below), it might be worth defining one or two metrics classes and then just having a train and test instance of a given class.

Reply:
I have now defined quite a few classes to handle this